proposals/reference-types/Overview.md

```diff
@@ -0,0 +1,204 @@
+
+* Set the stage for later additions:
+
+  - Typed function references (see below)
+  - Exception references (see exception handling proposal)
+  - A smoother transition path to GC (see GC proposal)
+
+Get the most important parts soon!
+
+Summary:
+
+* Add a new type `anyref` that can be used as both a value type and a table element type.
+
+* Also allow `anyfunc` as a value type.
+
+* Introduce instructions to get and set table slots.
+
+* Allow multiple tables.
+
+Notes:
+
+* This extension does not imply GC by itself, only if host refs are GCed pointers!
+
+* Reference types are *opaque*, i.e., their value is abstract and they cannot be stored into linear memory. Tables are used as the equivalent.
+
+
+## Language Extensions
+
+Typing extensions:
+
+* Introduce `anyref`, `anyfunc`, and `nullref` as a new class of *reference types*.
+  - `reftype ::= anyref | anyfunc | nullref`
+
+* Value types (of locals, globals, function parameters and results) can now be either numeric types or reference types.
+  - `numtype ::= i32 | i64 | f32 | f64`
+  - `valtype ::= <numtype> | <reftype>`
+  - locals with reference type are initialised with `null`
+
+* Element types (of tables) are equated with reference types.
+  - `elemtype ::= <reftype>`
+
+* Introduce a simple subtype relation between reference types.
+  - `nullref < t` for all reftypes `t`
+  - `t < anyref` for all reftypes `t`
+  - `anyfunc < anyref`
+
+
+New/extended instructions:
+
+* The new instruction `ref.null` evaluates to the null reference constant.
+  - `ref.null : [] -> [nullref]`
+
+* The new instructions `table.get` and `table.set` access tables.
+  - `table.get $x : [i32] -> [t]` iff `t` is the element type of table `$x`
+  - `table.set $x : [i32 t] -> []` iff `t` is the element type of table `$x`
+  - `table.fill $x : [i32 i32 t] -> []` iff `t` is the element type of table `$x`
+
+* The `call_indirect` instruction takes an additional table index as immediate that identifies the table it calls through.
+  - `call_indirect (type $t) $x : [t1* i32] -> [t2*]` iff `$t` denotes the function type `[t1*] -> [t2*]` and the element type of table `$x` is a subtype of `anyfunc`.
+  - In the binary format, space for the index is already reserved.
+  - For backwards compatibility, the index may be omitted in the text format, in which case it defaults to 0.
+
+
+Table extensions:
+
+* A module may define, import, and export multiple tables.
+  - As usual, the imports come first in the index space.
+  - This is already representable in the binary format.
+
+* Element segments take a table index as immediate that identifies the table they apply to.
+  - In the binary format, space for the index is already reserved.
+  - For backwards compatibility, the index may be omitted in the text format, in which case it defaults to 0.
+
+
+API extensions:
+
+* Any JS object (non-primitive value) or `null` can be passed as `anyref` to a Wasm function, stored in a global, or in a table.
```

Comment:
It would be good to add a TODO to figure out how strings fit in.

Reply:
Done.

Reply:
Added a subsection on equality and respective notes based on offline discussion -- for now leaving it as a possible extension, since it's not strictly needed and has some repercussions on type imports, even though we will probably want to include it anyway.